src/packagedcode/models.py

```diff
@@ -1304,6 +1304,13 @@ def populate_license_fields(cls, package_data):
         if package_data.extracted_license_statement and not isinstance(package_data.extracted_license_statement, str):
             package_data.extracted_license_statement = repr(package_data.extracted_license_statement)
 
+    @classmethod
+    def get_top_level_resources(cls, manifest_resource, codebase):
+        """
+        Yield Resources that are considered top-level for a Package type.
+        """
+        pass
```

Comment:
Should we have some reasonable default here, which for source archives of most ecosystems is files that are actually on the top level (and then this would be overridden like in the case of Maven jars, the function you added)? 
Or should this be in a different function which can be called in individual `get_top_level_resources` functions and in this way we don't have any reasonable defaults (and thus more accurate atleast, support for key files for various datafile handlers can be added later).

What would be better in the summary plugin context?


I was also looking to use these functions for getting key files (I will add another filter of `get_legal_or_readme()` function as used currently like [here](https://github.com/nexB/scancode-toolkit/blob/package_key_files/src/packagedcode/licensing.py#L348) to get selective key files which usually has license info from all the top level files) in package license detection post processing:

If we don't have license detections in extracted license statements from a package manifest:

we currently get sibling files which are legalese, and add license detections present there to the package license detections and populate other license fields from there. Like [here](https://github.com/nexB/scancode-toolkit/blob/package_key_files/src/packagedcode/plugin_package.py#L192) 

Does this make sense to use `get_top_level_resources` functions added in this PR for these?

Reply:
@AyanSinhaMahapatra 

> What would be better in the summary plugin context?

Currently, the only piece of code calling the new `get_top_level_resources` method is in the file classification plugin, where when we encounter a Resource with package_data, we "reconstitute" the Package from the package data and then run `get_top_level_resources` (https://github.com/nexB/scancode-toolkit/blob/6a561109b3fc8362d3b47d7bf2fda34a6d6ea1eb/src/summarycode/classify_plugin.py#L136)

If we scan a codebase with no package information in it, then the file classification plugin works as before. I haven't checked this for myself, but I would think that the same behavior as the default behavior for `get_top_level_resources` like what you are described in your first paragraph.

If you wanted to use the `get_top_level_resources` method in other places, like for the parts you're working on, then I think that having a default implementation makes sense. My intentions for not implementing a default behavior was that I wanted to not change the way the file classification plugin works, except in the case of a JAR.

Reply:
@AyanSinhaMahapatra Though I'd like to get this idea merged in first, then we can implement `get_top_level_resources` for more packages